### PR TITLE
Create common GSOF library

### PR DESCRIPTION
**Pull request by Ryanf55**
# Purpose

Factor out the parser into its own library so we can use it in EARHS for the PX1. Do it with as little change to the code as possible. 

# Testing Done

SITL test flights in a few different modes on copter and plane. I've also tested on the bench hardware connected to SITL. and a flight test.

# What changed

95% of the code is the same as before, the only difference is that the number of expected packets (the `valid` count), is now exposed up to higher levels. The parser used to hard-code 5 different packet types, which breaks down in EARHS. 
Otherwise, the parser no longer depends directly on types in AP_GPS, so there are now a function in AP_GPS_GSOF called `pack_state_data` to go from the packed GSOF data into the AP_GPS struct.

Longer term, I'd like to switch to a bitfield on which packets were received per DCOL frame.

# Next up

Add the EARHS. To see the WIP branch, see here: https://github.com/Ryanf55/ardupilot/tree/27033-gsof-add-eahrs-driver

# Issue

* Relates to #27033

### CODE REVIEW

**Comment by Ryanf55** [comment]
Flight tests complete.

File: libraries/AP_GPS/AP_GPS_GSOF.cpp [hunk 1]
```diff
@@ -373,4 +167,49 @@ AP_GPS_GSOF::validate_com_port(const uint8_t com_port) const {
     }
 }
 
+void
+AP_GPS_GSOF::pack_state_data() {
```
Comment (rmackay9) [inline libraries/AP_GPS/AP_GPS_GSOF.cpp:171]: nitpick: normally in cpp files we put the curly brace on a new line.

File: libraries/AP_GSOF/AP_GSOF.cpp [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    uint32_t u;
+    memcpy(&u, &src[pos], sizeof(u));
+    return be32toh(u);
+}
+
+uint16_t
+AP_GSOF::SwapUint16(const uint8_t* src, const uint32_t pos) const
+{
+    uint16_t u;
+    memcpy(&u, &src[pos], sizeof(u));
```
Comment (tridge) [inline libraries/AP_GSOF/AP_GSOF.cpp:133]: should use be16toh_ptr()

File: libraries/AP_GPS/AP_GPS_GSOF.cpp [hunk 3]
```diff
@@ -373,4 +167,49 @@ AP_GPS_GSOF::validate_com_port(const uint8_t com_port) const {
...
+    } else {
+        state.status = AP_GPS::NO_FIX;
+    }
+
+    state.location.lat = (int32_t)(RAD_TO_DEG_DOUBLE * position.latitude * (double)1e7);
+    state.location.lng = (int32_t)(RAD_TO_DEG_DOUBLE * position.longitude * (double)1e7);
+    state.location.alt = (int32_t)(position.altitude * 100);
+    state.last_gps_time_ms = AP_HAL::millis();
+
+    if ((vel.velocity_flags & 1) == 1)
```
Comment (rmackay9) [inline libraries/AP_GPS/AP_GPS_GSOF.cpp:198]: nitpick: although here we would normally put the curly brace on the same line as the if.

File: libraries/AP_GSOF/AP_GSOF.cpp [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    doubleu.bytes[6] = src[pos + 1];
+    doubleu.bytes[7] = src[pos + 0];
+
+    return doubleu.d;
+}
+
+float
+AP_GSOF::SwapFloat(const uint8_t* src, const uint32_t pos) const
+{
+    union {
```
Comment (tridge) [inline libraries/AP_GSOF/AP_GSOF.cpp:109]: please use be32tofloat_ptr()

File: libraries/AP_GSOF/AP_GSOF.cpp [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        break;
+    }
+
+    return NO_GSOF_DATA;
+}
+
+double
+AP_GSOF::SwapDouble(const uint8_t* src, const uint32_t pos) const
+{
+    union {
```
Comment (tridge) [inline libraries/AP_GSOF/AP_GSOF.cpp:90]: be64todouble_ptr()

File: libraries/AP_GSOF/AP_GSOF.cpp [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+#include "AP_GSOF.h"
+
+#if AP_GSOF_ENABLED
+
+
+
+#include <AP_Logger/AP_Logger.h>
+#include <AP_HAL/utility/sparse-endian.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_GSOF/AP_GSOF.h>
```
Comment (rmackay9) [inline libraries/AP_GSOF/AP_GSOF.cpp:13]: nitpick: this is a redundant include because "AP_GSOF.h" is above.

**Review by tridge — DISMISSED** [review]

File: libraries/AP_GSOF/AP_GSOF.cpp [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        msg.endtx = temp;
+        msg.state = Msg_Parser::State::STARTTX;
+        break;
+    }
+
+    return NO_GSOF_DATA;
+}
+
+double
+AP_GSOF::SwapDouble(const uint8_t* src, const uint32_t pos) const
```
Comment (rmackay9) [inline libraries/AP_GSOF/AP_GSOF.cpp:88]: I know it's existing code but we normally done use CamelCase... or maybe it's Pascal case.

Comment (tridge) [reply libraries/AP_GSOF/AP_GSOF.cpp:88]: or remove the function

File: libraries/AP_GSOF/AP_GSOF.h [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        uint8_t pos_flags2;
+        uint8_t initialized_number;
+    };
+
+    pos_time_t pos_time;
+
+    // GSOF2
+    struct PACKED pos_t {
+        // [rad]
+        double latitude;
```
Comment (rmackay9) [inline libraries/AP_GSOF/AP_GSOF.h:47]: latitude expressed in radians! I'd definitely put a _rad on the end of the variable name.

**Comment by Ryanf55** [comment]
Valgrind results: 
```
$ valgrind --soname-synonyms=somalloc=nouserintercepts ./build/linux/tests/test_gsof 
==46624== Memcheck, a memory error detector
==46624== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==46624== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==46624== Command: ./build/linux/tests/test_gsof
==46624== 
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from AP_GSOF
[ RUN      ] AP_GSOF.incomplete_packet
[       OK ] AP_GSOF.incomplete_packet (5 ms)
[ RUN      ] AP_GSOF.packet1
[       OK ] AP_GSOF.packet1 (9 ms)
[----------] 2 tests from AP_GSOF (17 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (39 ms total)
[  PASSED  ] 2 tests.
==46624== 
==46624== HEAP SUMMARY:
==46624==     in use at exit: 0 bytes in 0 blocks
==46624==   total heap usage: 222 allocs, 222 frees, 118,093 bytes allocated
==46624== 
==46624== All heap blocks were freed -- no leaks are possible
==46624== 
==46624== For lists of detected and suppressed errors, rerun with: -s
==46624== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

File: libraries/AP_GPS/AP_GPS_GSOF.cpp (hunk 3 above)
Comment (Ryanf55) [reply libraries/AP_GPS/AP_GPS_GSOF.cpp:198]: I'm using ArduPilot's Astyle formatter on the code, and encorcing it in CI. Do you know how to get the formatter to fix this?

Comment (Ryanf55) [reply libraries/AP_GPS/AP_GPS_GSOF.cpp:198]: Oh nevermind, this is the GPS, which is not autoformatted.

**Comment by Ryanf55** [comment]
Valgrind doesn't work on the autotest right now:
```
 ./Tools/autotest/autotest.py build.Copter  test.CopterTests2b.GPSTypes --valgrind
 ...
 arducopter: valgrind: Startup or configuration error:
arducopter: valgrind:    Can't create client cmdline file in /home/ryan/Dev/ardu_ws/src/ardupilot/tmp/valgrind_proc_70431_cmdline_471259db
arducopter: valgrind: Unable to start up properly.  Giving up.
>>>> FAILED STEP: test.CopterTests2b.GPSTypes at Mon May 27 18:29:49 2024 (End Of File (EOF). Exception style platform.
<pexpect.pty_spawn.spawn object at 0x7f23d9dcdd20>
command: /usr/bin/valgrind
args: [b'/usr/bin/valgrind', b'--soname-synonyms=somalloc=nouserintercepts', b'--vgdb-prefix=/tmp/vgdb-pipe', b'-q', b'--log-file=arducopter-+-valgrind.log', b'/home/ryan/Dev/ardu_ws/src/ardupilot/build/sitl/bin/arducopter', b'-w', b'-S', b'--home', b'-35.362938,149.165085,584,270', b'--model', b'+', b'--serial1=tcp:2', b'--defaults', b'/tmp/tmpjmreei1d,/home/ryan/Dev/ardu_ws/src/ardupilot/Tools/autotest/default_params/copter.parm']
buffer (last 100 chars): ''
before (last 100 chars): 'pilot/tmp/valgrind_proc_70431_cmdline_471259db\r\nvalgrind: Unable to start up properly.  Giving up.\r\n'
after: <class 'pexpect.exceptions.EOF'>
match: None
match_index: None
exitstatus: None
flag_eof: True
pid: 70431
child_fd: 6
closed: False
timeout: 5
delimiter: <class 'pexpect.exceptions.EOF'>
logfile: <pysim.util.PSpawnStdPrettyPrinter object at 0x7f23d9dce080>
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
searcher: searcher_re:
    0: re.compile('Waiting for '))
Traceback (most recent call last):
  File "/home/ryan/Dev/ardu_ws/src/ardupilot/./Tools/autotest/autotest.py", line 719, in run_tests
    success = run_step(step)
  File "/home/ryan/Dev/ardu_ws/src/ardupilot/./Tools/autotest/autotest.py", line 526, in run_step
    return run_specific_test(specific_test_to_run, binary, **fly_opts)
  File "/home/ryan/Dev/ardu_ws/src/ardupilot/./Tools/autotest/autotest.py", line 387, in run_specific_test
    return tester.autotest(tests=[a], allow_skips=False, step_name=step), tester
  File "/home/ryan/Dev/ardu_ws/src/ardupilot/Tools/autotest/vehicle_test_suite.py", line 14091, in autotest
    results = self.run_tests(tests)
  File "/home/ryan/Dev/ardu_ws/src/ardupilot/Tools/autotest/vehicle_test_suite.py", line 11780, in run_tests
    self.init()
  File "/home/ryan/Dev/ardu_ws/src/ardupilot/Tools/autotest/vehicle_test_suite.py", line 8738, in init
    self.start_SITL()
  File "/home/ryan/Dev/ardu_ws/src/ardupilot/Tools/autotest/vehicle_test_suite.py", line 8652, in start_SITL
    self.sitl = util.start_SITL(binary, **start_sitl_args)
  File "/home/ryan/Dev/ardu_ws/src/ardupilot/Tools/autotest/pysim/util.py", line 608, in start_SITL
    child.expect('Waiting for ', timeout=300)
  File "/home/ryan/.local/lib/python3.10/site-packages/pexpect/spawnbase.py", line 354, in expect
    return self.expect_list(compiled_pattern_list,
  File "/home/ryan/.local/lib/python3.10/site-packages/pexpect/spawnbase.py", line 383, in expect_list
    return exp.expect_loop(timeout)
  File "/home/ryan/.local/lib/python3.10/site-packages/pexpect/expect.py", line 179, in expect_loop
    return self.eof(e)
  File "/home/ryan/.local/lib/python3.10/site-packages/pexpect/expect.py", line 122, in eof
    raise exc
pexpect.exceptions.EOF: End Of File (EOF). Exception style platform.
<pexpect.pty_spawn.spawn object at 0x7f23d9dcdd20>
command: /usr/bin/valgrind
args: [b'/usr/bin/valgrind', b'--soname-synonyms=somalloc=nouserintercepts', b'--vgdb-prefix=/tmp/vgdb-pipe', b'-q', b'--log-file=arducopter-+-valgrind.log', b'/home/ryan/Dev/ardu_ws/src/ardupilot/build/sitl/bin/arducopter', b'-w', b'-S', b'--home', b'-35.362938,149.165085,584,270', b'--model', b'+', b'--serial1=tcp:2', b'--defaults', b'/tmp/tmpjmreei1d,/home/ryan/Dev/ardu_ws/src/ardupilot/Tools/autotest/default_params/copter.parm']
buffer (last 100 chars): ''
before (last 100 chars): 'pilot/tmp/valgrind_proc_70431_cmdline_471259db\r\nvalgrind: Unable to start up properly.  Giving up.\r\n'
after: <class 'pexpect.exceptions.EOF'>
match: None
match_index: None
exitstatus: None
flag_eof: True
pid: 70431
child_fd: 6
closed: False
timeout: 5
delimiter: <class 'pexpect.exceptions.EOF'>
logfile: <pysim.util.PSpawnStdPrettyPrinter object at 0x7f23d9dce080>
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
searcher: searcher_re:
    0: re.compile('Waiting for ')
FAILED 1 tests: ['test.CopterTests2b.GPSTypes']
```

File: libraries/AP_GSOF/AP_GSOF.h (hunk 8 above)
Comment (Ryanf55) [reply libraries/AP_GSOF/AP_GSOF.h:47]: Done

**Comment by Ryanf55** [comment]
@peterbarker I am still facing the valgrind problem after rebasing. Can you try running the test on your system?

File: libraries/AP_GSOF/AP_GSOF.cpp [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    floatu.bytes[2] = src[pos + 1];
+    floatu.bytes[3] = src[pos + 0];
+
+    return floatu.f;
+}
+
+uint32_t
+AP_GSOF::SwapUint32(const uint8_t* src, const uint32_t pos) const
+{
+    uint32_t u;
```
[the remark this answers is not in the document]
Comment (Ryanf55) [reply libraries/AP_GSOF/AP_GSOF.cpp:123]: If it's ok, I plan to convert all these in a follow up PR. This was moving the code to a library in an attempt not to change anything.

**Comment by Ryanf55** [comment]
I just rebased on master, the valgrind test doesn't work still. Considering this was already flight tested successfully, can we merge it? AP_GSOF doesn't call `new`.

**Comment by Ryanf55** [comment]
Need to fix CI: 
```
    OUT: [----------] 2 tests from AP_GSOF
    OUT: [ RUN      ] AP_GSOF.incomplete_packet
    OUT: [       OK ] AP_GSOF.incomplete_packet (0 ms)
    OUT: [ RUN      ] AP_GSOF.packet1
    OUT: ../../libraries/AP_GSOF/tests/test_gsof.cpp:25: Failure
    OUT: Expected: (fp) != (nullptr), actual: NULL vs (nullptr)
check: 1 of 49 tests failed
    /__w/ardupilot/ardupilot/build/linux/tests/test_gsof
 ```

**Comment by Ryanf55** [comment]
> Need to fix CI:
> 
> ```
>     OUT: [----------] 2 tests from AP_GSOF
>     OUT: [ RUN      ] AP_GSOF.incomplete_packet
>     OUT: [       OK ] AP_GSOF.incomplete_packet (0 ms)
>     OUT: [ RUN      ] AP_GSOF.packet1
>     OUT: ../../libraries/AP_GSOF/tests/test_gsof.cpp:25: Failure
>     OUT: Expected: (fp) != (nullptr), actual: NULL vs (nullptr)
> check: 1 of 49 tests failed
>     /__w/ardupilot/ardupilot/build/linux/tests/test_gsof
> ```

This works:
```
waf tests
build/sitl/tests/test_gsof
```

This doesn't. 
```
CI_BUILD_TARGET="linux" ./Tools/scripts/build_ci.sh 
```

Until we have a better option for loading in a data file, I'm disabling the binary file test.

**Comment by tridge** [comment]
@Ryanf55 where are we up to with this?

**Comment by Ryanf55** [comment]
> @Ryanf55 where are we up to with this?

I've met the requests of last dev call from Randy. The changes of functions to use the functions in spare-endian is ready for a follow PR.

This has been flown and has unit tests. 

Therefore, I would like to request a merge.

**Review by tridge — APPROVED** [review]

**Pull request merged**